src/mbgl/map/map.cpp

```diff
@@ -276,46 +249,70 @@ void Map::Impl::update() {
 
     style->updateTiles(parameters);
 
-    if (mode == MapMode::Continuous) {
-        backend.invalidate();
-    } else if (stillImageRequest && style->isLoaded()) {
-        // TODO: determine whether we need activate/deactivate
-        BackendScope guard(backend);
-        render(stillImageRequest->view);
-    }
-
     updateFlags = Update::Nothing;
-}
 
-void Map::Impl::render(View& view) {
     if (!painter) {
         painter = std::make_unique<Painter>(backend.getContext(), transform.getState(), pixelRatio);
     }
 
-    FrameData frameData { timePoint,
-                          pixelRatio,
-                          mode,
-                          contextMode,
-                          debugOptions };
+    if (mode == MapMode::Continuous) {
+        if (renderState == RenderState::Never) {
+            backend.notifyMapChange(MapChangeWillStartRenderingMap);
+        }
 
-    painter->render(*style,
-                    frameData,
-                    view,
-                    annotationManager->getSpriteAtlas());
+        backend.notifyMapChange(MapChangeWillStartRenderingFrame);
+
+        FrameData frameData { timePoint,
+                              pixelRatio,
+                              mode,
+                              contextMode,
+                              debugOptions };
+
+        painter->render(*style,
+                        frameData,
+                        view,
+                        annotationManager->getSpriteAtlas());
+
+        painter->cleanup();
+
+        backend.notifyMapChange(style->isLoaded() ?
+            MapChangeDidFinishRenderingFrameFullyRendered :
+            MapChangeDidFinishRenderingFrame);
+
+        if (!style->isLoaded()) {
+            renderState = RenderState::Partial;
+        } else if (renderState != RenderState::Fully) {
+            renderState = RenderState::Fully;
+            backend.notifyMapChange(MapChangeDidFinishRenderingMapFullyRendered);
+            if (loading) {
+                loading = false;
+                backend.notifyMapChange(MapChangeDidFinishLoadingMap);
+            }
+        }
+
+        if (style->hasTransitions()) {
+            onUpdate(Update::RecalculateStyle);
```

Comment:
In this and the `painter->needsAnimation()` case, should we be `|`'ing `Update::RecalculateStyle` / `Update::Repaint` with `flags` (from `updateTransitions`)?